presto-cli/src/main/java/com/facebook/presto/cli/AbstractWarningsPrinter.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.cli;
+
+import com.facebook.presto.spi.PrestoWarning;
+import com.google.common.collect.AbstractIterator;
+
+import java.util.Iterator;
+import java.util.List;
+
+import static com.facebook.presto.cli.ConsolePrinter.REAL_TERMINAL;
+import static java.lang.String.format;
+
+abstract class AbstractWarningsPrinter
+        implements WarningsPrinter
+{
+    private static final String WARNING_BEGIN = ((char) 27) + "[33m";
+    private static final String WARNING_END = ((char) 27) + "[39m";
+
+    private boolean hasProcessedWarnings;
+    private int processedWarnings;
+
+    protected String getWarningMessage(PrestoWarning warning)
+    {
+        // If this is a real terminal color the warnings yellow
+        return REAL_TERMINAL ? format("%sWARNING: %s%s", WARNING_BEGIN, warning.getMessage(), WARNING_END) :
+                format("WARNING: %s", warning.getMessage());
+    }
+
+    protected int getProcessedWarnings()
+    {
+        return processedWarnings;
+    }
+
+    protected Iterator<PrestoWarning> warningIterator(List<PrestoWarning> warnings)
+    {
+        return new AbstractIterator<PrestoWarning>() {
+            @Override
+            protected PrestoWarning computeNext()
+            {
+                if (processedWarnings < warnings.size()) {
+                    return warnings.get(processedWarnings++);
```

Comment:
Having a side-effect like this hidden in this method is error prone and makes it harder to reason about the code. Anyone who reads code that calls `warningIterator` is going to miss this fact and also wonder why there's an explicit method instead of calling `warnings.iterator()`. Let me think of a better way to structure this...

Reply:
Makes sense, updated with a different approach: removed the iterator and moved as much logic as possible to the AbstractWarningsPrinter. Let me know if you are ok with this approach.